### PR TITLE
[C++] Improve C++ Cookbook

### PR DESCRIPTION
**Pull request by js8544**
ARROW-17851
Add ARROW_ASSIGN_OR_RAISE to utility macros.
Use auto when it improves readability.
Adopt C++ 14/17's idioms.
Add Visitor classes to visitor pattern.

### CODE REVIEW

File: cpp/code/basic_arrow.cc [hunk 1]
```diff
@@ -70,48 +76,48 @@ TEST(BasicArrow, ReturnNotOk) { ASSERT_OK(ReturnNotOk()); }
...
+      return arrow::Status::NotImplemented("Can not compute sum for array of type ",
+                                           array.type()->ToString());
     }
-    return arrow::Status::OK();
   }
 };  // TableSummation
 
 arrow::Status VisitorSummationExample() {
   StartRecipe("VisitorSummationExample");
-  std::shared_ptr<arrow::Schema> schema = arrow::schema({
```
Comment (lidavidm) [inline cpp/code/basic_arrow.cc:101]: Note that in general, it was consciously chosen to write out return types to be explicit for readers who may be unfamiliar with Arrow. Some things like lambdas and initializing a builder on the stack make sense to simplify to me, but helpers like `schema` should be written out.

File: cpp/code/basic_arrow.cc [hunk 2]
```diff
@@ -35,6 +35,11 @@ arrow::Status ReturnNotOkMacro() {
     if (!st.ok()) {
       return st;
     }
+    auto res = builder.Finish();
+    std::shared_ptr<arrow::Array> arr;
+    if (res.ok()) {
+      arr = std::move(res).ValueUnsafe();
+    }
```
Comment (lidavidm) [inline cpp/code/basic_arrow.cc:42]: This recipe is specifically trying to demonstrate how `Status` works and not how to construct an Array, really.

File: cpp/source/basic.rst [hunk 3]
```diff
@@ -41,9 +41,9 @@ tedious:
   :caption: Checking the status of every function manually
   :dedent: 2
 
-The macro :c:macro:`ARROW_RETURN_NOT_OK` will take care of some of this
-boilerplate for you.  It will run the contained expression and check the resulting
-``Status`` or ``Result`` object.  If it failed then it will return the failure.
+The macros :c:macro:`ARROW_RETURN_NOT_OK` and :c:macro:`ARROW_ASSIGN_OR_RAISE` will take care of some of these
+boilerplates for you.  They will run the contained expression and check the resulting
+``Status`` or ``Result`` object.  If they failed then they will return the failure.
```
Comment (lidavidm) [inline cpp/source/basic.rst:46]: For clarity, I think Status and Result should be introduced separately

File: cpp/code/basic_arrow.cc (hunk 2 above)
Comment (js8544) [reply cpp/code/basic_arrow.cc:42]: Hmm, the section that refers this code snippet is titled "Working with Status and Result" but there's no usage of `Result` at all. That's why I added these lines to demonstrate the usage of Result.

File: cpp/code/basic_arrow.cc (hunk 1 above)
Comment (js8544) [reply cpp/code/basic_arrow.cc]: I see. I've reverted most `auto`s in f069a5be4d6544e26edd6735823e7d0442c87564

File: cpp/source/basic.rst (hunk 3 above)
Comment (js8544) [reply cpp/source/basic.rst:46]: Sure. Should I split the section "Working with Status and Result" into two or should I keep it but move Result related stuff to a separate paragraph?

Comment (lidavidm) [reply cpp/source/basic.rst:46]: I think one section is OK, just have the text separate so it's clear what applies to what

Comment (lidavidm) [reply cpp/source/basic.rst:46]: (And separate code blocks)

File: cpp/code/basic_arrow.cc (hunk 2 above)
Comment (lidavidm) [reply cpp/code/basic_arrow.cc]: As much as possible I think we should try to keep each code sample focused on one thing and one thing only, hence why I'd rather not add it here. And ideally we'd have code both for this approach as well as for the macro-based approach.

File: cpp/code/creating_arrow_objects.cc [hunk 4]
```diff
@@ -28,7 +28,7 @@ arrow::Status CreatingArrays() {
   ARROW_RETURN_NOT_OK(builder.Append(1));
   ARROW_RETURN_NOT_OK(builder.Append(2));
   ARROW_RETURN_NOT_OK(builder.Append(3));
-  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::Array> arr, builder.Finish())
+  ARROW_ASSIGN_OR_RAISE(auto arr, builder.Finish())
```
Comment (lidavidm) [inline cpp/code/creating_arrow_objects.cc:31]: I'd probably want to revert these autos as well

**Review by lidavidm — COMMENTED** [review]
I think the other issue is technically we aren't switching to C++17 until Arrow 10 so we may not want to update the cookbook with that until release

**Comment by js8544** [comment]
> 

Oh right. I'll put this PR on hold for now.

**Comment by lidavidm** [comment]
This should now be OK if rebased

**Comment by js8544** [comment]
> This should now be OK if rebased

Thanks! I'll work on this sometime next week.

**Pull request closed without merging**